source/tutorials/kubernetes.rst

```diff
@@ -3,354 +3,716 @@
+
+
+Setting proxy servers for Kubernetes
+====================================
+
+If you receive any of the messages below, check outbound Internet access. You
+may be behind a proxy server. 
+
+   * Images cannot be pulled.
+   * Connection refused error.
+   * Connection timed-out or Access Refused errors.
+   * The warnings when :command:`kubeadm init` is run.
+     .. code-block:: console
+   
+        [WARNING HTTPProxy]: Connection to "https://<HOST-IP>" uses proxy "<PROXY-SERVER>". If that is not intended, adjust your proxy settings
+        [WARNING HTTPProxyCIDR]: connection to "10.96.0.0/12" uses proxy "<PROXY-SERVER>". This may lead to malfunctional cluster setup. Make sure that Pod and Services IP ranges specified correctly as exceptions in proxy configuration
+        [WARNING HTTPProxyCIDR]: connection to "10.244.0.0/16" uses proxy "<PROXY-SERVER>". This may lead to malfunctional cluster setup. Make sure that Pod and Services IP ranges specified correctly as exceptions in proxy configuration
+
+
+If you use an outbound proxy server, you must configure proxy settings
+appropriately for all components in the stack including :command:`kubectl` and
+container runtime services. 
+
+Configure the :ref:`proxy settings <proxy>`, using the standard *HTTP_PROXY*,
+*HTTPS_PROXY*, and *NO_PROXY* environment variables. The *NO_PROXY* values are
+especially important for Kubernetes to ensure private IP traffic does not try
+to go out the proxy.
+
+#. Set your environment proxy variables. Ensure that your local IP address is
+   **explicitly included** in the environment variable *NO_PROXY*. Setting
+   *localhost* is not sufficient!
+
+   .. code-block:: bash
+
+      export http_proxy=http://proxy.example.com:80
+      export https_proxy=http://proxy.example.com:443
+      export no_proxy=.svc,10.0.0.0/8,192.168.0.0/16,`hostname`,localhost
+
+   .. important::
+
+      :command:`kubeadm` commands specifically use these shell variables for proxy
+      configuration. Ensure they are set your running terminal before running
+      :command:`kubeadm` commands.
+
+#. Run the following command to add systemd drop-in configurations for each
+   service to include proxy settings:
+
+   .. code-block:: bash
+
+      services=(kubelet docker crio containerd)
+      for s in "${services[@]}"; do
+      sudo mkdir -p "/etc/systemd/system/${s}.service.d/"
+      cat << EOF | sudo tee "/etc/systemd/system/${s}.service.d/proxy.conf"
+      [Service]
+      Environment="HTTP_PROXY=${http_proxy}"
+      Environment="HTTPS_PROXY=${https_proxy}"
+      Environment="SOCKS_PROXY=${socks_proxy}"
+      Environment="NO_PROXY=${no_proxy}"
+      EOF
+      done
+
+
+#. Reload the systemd manager configuration.
+
+   .. code-block:: bash
+      
+      sudo systemctl daemon-reload
+
+
+If you had a previously failed initialization due to a proxy issue, restart
+the process with the :command:`kubeadm reset` command.
+
+
+DNS issues
```

Comment:
I found a few more error conditions and how to solve. These need to be wordsmith ed and extraneous details removed. 

1. coredns pods stuck in container creating.  Logs show something like the following.
```
  Warning  FailedCreatePodSandBox  5m7s                 kubelet, kata3     Failed to create pod sandbox: rpc error: code = Unknown desc = failed to get network JSON for pod sandbox k8s_coredns-6955765f44-5gpj2_kube-system_9ff6ddc2-189f-40fc-bd48-8a392dfdc744_0(ea3135094c88babb982f45de17fe5716a3320dd2eb79990c5eb8959c11cfc680): cannot convert version ["" "0.1.0" "0.2.0"] to 0.4.0
```
Soln: In this case the /etc/cni/net.d/10-flannel.conflist or another CNI file is using an incompatible (probably older) version. Delete the file and restart the stack. 

2. coredns pod stuck in container creating. Logs show something like the following.
```
  Warning  FailedCreatePodSandBox  117s (x197 over 45m)  kubelet, kata3     (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_coredns-6955765f44-npsm5_kube-system_323177a2-2577-4748-96ed-dd8c265d8ad4_0(98b34e90afd36c948c171977248313d5e378a8e1926a14377d922cb22db9f104): error getting ClusterInformation: Get https://[10.96.0.1]:443/apis/crd.projectcalico.org/v1/clusterinformations/default: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kubernetes")
```
soln: There are probably multiple CNI configuration files in /etc/cni/net.d  Delete them all and start again.

3. coredns pods stuck in pending
```
Warning  FailedScheduling  55s (x3 over 2m12s)  default-scheduler  0/1 nodes are available: 1 node(s) had taints that the pod didn't tolerate.
```
soln 2.1 - If /etc/cni/net.d is blank then try applying a CNI plugin. It should populate that directory with a CNI configuration

Reply:
I added these to the PR.